crates/scheduler/src/schedule/mod.rs

```diff
@@ -0,0 +1,910 @@
+//! Schedules that produce orderings of systems that are correct.
+//!
+//! Correct, meaning they (try to) guarantee:
+//! * freedom from system starvation, meaning all systems get to execute in the schedule ordering;
+//! * freedom from race conditions, meaning the ordering will not place reads and writes to
+//!   the same component at the same time;
+//! * freedom from deadlock, meaning systems that are ordered such that they will always
+//!   be able to progress.
+
+use crate::precedence::{Orderable, Precedence};
+use crate::schedule::PrecedenceGraphError::{
+    Deadlock, Dependency, IncorrectSystemCompletionMessage, PendingSystemIndexNotFound,
+};
+use crossbeam::channel::{Receiver, RecvError, Select};
+use daggy::petgraph::visit::{IntoNeighbors, IntoNeighborsDirected, IntoNodeIdentifiers};
+use daggy::petgraph::{visit, Incoming};
+use daggy::{Dag, NodeIndex, WouldCycle};
+use ecs::{Schedule, ScheduleError, ScheduleResult, System, SystemExecutionGuard};
+use itertools::Itertools;
+use std::fmt::{Debug, Display, Formatter};
+use thiserror::Error;
+use tracing::{debug, error};
+
+type Sys<'system> = &'system dyn System;
+type SysDag<'system> = Dag<Sys<'system>, i32>;
+
+/// An error occurred during a precedence graph schedule operation.
+#[derive(Error, Debug)]
+enum PrecedenceGraphError {
+    /// Dependency construction error.
+    #[error("dependency from `{from}` to `{to}` in graph {graph} would cause a cycle")]
+    Dependency {
+        /// The node the dependency is coming from.
+        from: String,
+        /// The node the dependency is going to.
+        to: String,
+        /// The current state of the graph.
+        graph: String,
+    },
+    /// A pending system was completed but there is no system at its index anymore.
+    #[error("a pending system was completed but there is no system at its index anymore")]
+    PendingSystemIndexNotFound(usize),
+    /// Expected the `SystemExecutionGuard` channel to be dropped to signal completion, but got message.
+    #[error("expected the `SystemExecutionGuard` channel to be dropped to signal completion, but got message {0:?}")]
+    IncorrectSystemCompletionMessage(Result<(), RecvError>),
+    /// The schedule has deadlocked due to there being no pending systems but all systems have still not run.
+    #[error("the schedule has deadlocked due to there being no pending systems but all systems have still not run")]
+    Deadlock,
+}
+
+/// Whether a precedence graph operation succeeded.
+type PrecedenceGraphResult<T, E = PrecedenceGraphError> = Result<T, E>;
+
+/// Orders [`System`]s based on their precedence.
+#[derive(Default, Clone)]
+pub struct PrecedenceGraph<'systems> {
+    dag: SysDag<'systems>,
+    /// Systems which have been given out, and are awaiting execution.
+    ///
+    /// Warning: Node indices are _not_ stable and will be invalidated if [`dag`] is mutated.
```

Comment:
The intra-doc link ``[`dag`]`` is broken because `dag` is not in scope. This can be fixed by replacing it to ``[`dag`](Self::dag)``.

Reply:
Neither seem to work for me, so I assumed it was impossible to have it correctly link. I'm only trying it inside CLion though, do you mean that `[`dag`](Self::dag)` would correctly link it if we were to generate out the documentation?

Reply:
When using ``[`dag`](Self::dag)`` the warning is gone when running `cargo doc`, but it looks like CLion still can't follow the link.

Reply:
Ah, I usually don't run `cargo doc` but that's good to know! I'll update it then for the sake of being correct. 👍